ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,526 @@
+   */
+  const getMessageWithContent = (
+    messageFromDb: RetrievedMessageWithoutContent
+  ): TaskEither<ActivityResultQueryFailure, MessageWithContent> => {
+    return fromQueryEither(
+      () => messageModel.getContentFromBlob(blobService, messageFromDb.id),
+      "messageModel.getContentFromBlob"
+    ).foldTaskEither<ActivityResultQueryFailure, MessageWithContent>(
+      failure => fromEither(left(failure)),
+      maybeContent =>
+        fromEither<ActivityResultQueryFailure, MessageContent>(
+          fromOption(
+            ActivityResultQueryFailure.encode({
+              kind: "QUERY_FAILURE",
+              query: "messageModel.getContentFromBlob",
+              reason: `Cannot find content for message ${messageFromDb.id}`
+            })
+          )(maybeContent)
+        ).map<MessageWithContent>((content: MessageContent) =>
+          appendContentToMessage(messageFromDb, content)
+        )
+    );
+  };
+
+  /**
+   * Utility that performs enrichMessage over a list of messages
+   * @param messages a list of messages without content
+   * @returns either a list of message with content or a query error
+   */
+  const getAllMessagesWithContent = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageWithContent>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
+      const [firstMessage, ...otherMessages] = messages;
+      return sequenceT(taskEither)(
+        getMessageWithContent(firstMessage),
+        ...otherMessages.map(getMessageWithContent)
+      );
+    }
+    return taskEither.of([]);
+  };
+
+  /**
+   * Given a list of messages, it queires for relative notifications
+   * @param messages
+   */
+  const findNotificationsForAllMessages = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<RetrievedNotification>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
+      const [firstMessage, ...otherMessages] = messages;
+
+      return sequenceT(taskEither)(
+        fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+          () =>
+            iteratorToArray(
+              notificationModel.findNotificationsForMessage(firstMessage.id)
+            ),
+          "findNotificationsForRecipient"
+        ),
+        ...otherMessages.map(m =>
+          fromQueryEither<ReadonlyArray<RetrievedNotification>>(
+            () =>
+              iteratorToArray(
+                notificationModel.findNotificationsForMessage(m.id)
+              ),
+            "findNotificationsForRecipient"
+          )
+        )
+      ).foldTaskEither(
+        e => fromEither(left(e)),
+        arrayOfArray =>
+          fromEither(
+            right(
+              arrayOfArray.reduce(
```

Comment:
you can use array.flatten from fp-ts

Reply:
I think it doesn't support readonly arrays (introduced in v2.5)